### PR TITLE
Performance improvement of `SparsePauliOp.to_matrix`

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

Performance improvement of `SparsePauliOp.to_matrix`.

I used Jake's technique https://github.com/Qiskit/qiskit-terra/pull/9229#pullrequestreview-1202481735.

```python
import random
from timeit import timeit

from qiskit.quantum_info import SparsePauliOp

k = 10
n = 1000

random.seed(123)
op = SparsePauliOp.from_list([(''.join(random.choices('IXYZ', k=k)), 1) for _ in range(n)])

print(f'dense {timeit(lambda: op.to_matrix(sparse=False), number=1)} sec')
print(f'sparse {timeit(lambda: op.to_matrix(sparse=True), number=1)} sec')
```

main
```
dense 9.802448930000082 sec
sparse 3.588473711000006 sec
```

this PR
```
dense 6.786199189000399 sec
sparse 3.1803273659998013 sec
```

### Details and comments

(benchmark for sparse)
```python
import random
from timeit import timeit

from qiskit.quantum_info import SparsePauliOp

k = 25
n = 1

random.seed(123)
op = SparsePauliOp.from_list([(''.join(random.choices('IXYZ', k=k)), 1) for _ in range(n)])

print(f'sparse {timeit(lambda: op.to_matrix(sparse=True), number=1)} sec')
```

main
```
sparse 17.364562793000005 sec
```

this PR
```
sparse 1.894934757000101 sec
```

(further optimization)
If I'm allowed to extend `BasePauli._to_matrix` to multiply `coeff` directly, we can have more improvement for the dense case because Pauli matrices are sparse. See https://github.com/t-imamichi/qiskit-terra/compare/opt-tomatrix2...t-imamichi:qiskit-terra:opt-tomatrix3 for the code diff.

opt-tomatrix3
```
dense 4.616774103999887 sec
sparse 3.021996235000188 sec
```

### CODE REVIEW

File: qiskit/quantum_info/operators/symplectic/base_pauli.py [hunk 1]
```diff
@@ -441,8 +445,7 @@ def _to_matrix(z, x, phase=0, group_phase=False, sparse=False):
 
         # Build dense matrix using csr format
         mat = np.zeros((dim, dim), dtype=complex)
-        for i in range(dim):
-            mat[i][indices[indptr[i] : indptr[i + 1]]] = data[indptr[i] : indptr[i + 1]]
+        mat[range(dim), indices[:dim]] = data[:dim]
```
Comment (t-imamichi) [inline qiskit/quantum_info/operators/symplectic/base_pauli.py:448]: This simplification is based on `indptr = np.arange(dim + 1, dtype=np.uint)`.

`indptr[i] : indptr[i+1]` -> `i: i+1`
`indices[i: i+1]` -> `indices[i]`
`data[i: i+1]` -> `data[i]`

File: qiskit/quantum_info/operators/symplectic/base_pauli.py [hunk 2]
```diff
@@ -432,7 +436,7 @@ def _to_matrix(z, x, phase=0, group_phase=False, sparse=False):
             coeff = (-1j) ** phase
         else:
             coeff = 1
-        data = np.array([coeff * (-1) ** (bin(i).count("1") % 2) for i in z_indices & indptr])
+        data = coeff * _PARITY[(z_indices & indptr).astype(np.uint8)]
```
Comment (ikkoham) [inline qiskit/quantum_info/operators/symplectic/base_pauli.py:439]: Why is uint8 enough? Is it possible to have a larger number?

**Review by ikkoham — COMMENTED** [review]
Thank you. Great performance improvement.

In actual use, we often want to be sparse for large qubits, so further optimization is not necessary at this time.

File: qiskit/quantum_info/operators/symplectic/base_pauli.py (hunk 2 above)
Comment (t-imamichi) [reply qiskit/quantum_info/operators/symplectic/base_pauli.py:439]: The original code computes the parity of each value of `z_indices & indptr`.
My code first applies `astype(np.uint8)` to take mod 256 (parity does not change) and then applies a lookup table `_PARITY[256]` to compute the parity. Does it make sense?

Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/base_pauli.py:439]: Thanks for your explanation. I misunderstood the definition of parity.

Comment (t-imamichi) [reply qiskit/quantum_info/operators/symplectic/base_pauli.py:439]: Sorry. I noticed that I misunderstood the party of the code. This code does not work with many qubits. I will rethink of the optimization.

Comment (t-imamichi) [reply qiskit/quantum_info/operators/symplectic/base_pauli.py]: fixed at fb1428c

**Comment by t-imamichi** [comment]
I added `test_to_matrix_large` to test `to_matrix` method with 10 qubit examples.

**Review by ikkoham — APPROVED** [review]
Great performance improvement. Current code works correctly!

**Pull request merged**